Submsg and replies

ethanfrey opened the pull request:
Closes #421 

- [x] Add private method to call reply entry point
- [x] Wrote logic to dispatch submessages
- [x] Dispatch submessages from all (non-ibc) entrypoints (see note below)
- [x] Test submessage handling via reflect contract
  - [x] no gas limit, success
  - [x] no gas limit, failure
  - [x] no gas limit, out-of-gas panic
  - [x] gas limit, success
  - [x] gas limit, failure
  - [x] gas limit, consume all gas
  - [x] Instantiate Contract, get address from reply.Data

Note: we will need to update again after https://github.com/CosmWasm/cosmwasm/issues/822 to add support for submessages in the ibc callbacks

alpe commented on x/wasm/internal/keeper/keeper.go:
```diff
@@ -469,7 +473,48 @@ func (k Keeper) Sudo(ctx sdk.Context, contractAddress sdk.AccAddress, msg []byte
+// reply is only called from keeper internal functions (dispatchSubmessages) after processing the submessage
+// it
+func (k Keeper) reply(ctx sdk.Context, contractAddress sdk.AccAddress, reply wasmvmtypes.Reply) (*sdk.Result, error) {
+	// we can assume this is still in cache and not charge again for returning the reply. correct?
```
The LRU cache might be disabled via `--wasm.memory_cache_size` start parameter.
I have mixed feelings not charging anything or charging the full instantiation fee (again). Let's discuss in a bigger group.

ethanfrey replied:
Decision was to charge the same fee here (which is 0 if pinned).
We can consider some discount later

alpe commented on x/wasm/internal/keeper/keeper.go:
```diff
@@ -684,15 +729,134 @@ func (k Keeper) GetByteCode(ctx sdk.Context, codeID uint64) ([]byte, error) {
+}
+
+func (k Keeper) dispatchMsgWithGasLimit(ctx sdk.Context, contractAddr sdk.AccAddress, ibcPort string, msg wasmvmtypes.CosmosMsg, gasLimit uint64) (events []sdk.Event, data [][]byte, err error) {
+	// FIXME: do I need to wrap the BlockGasMeter in the same way?
```
IMHO the block gas meter should not be wrapped here. If we consume more than available for the block then it would fail later anyway: https://docs.cosmos.network/master/basics/gas-fees.html#block-gas-meter

ethanfrey replied:
Thanks, that was what I thought, but was unsure

alpe commented on x/wasm/internal/keeper/keeper.go:
```diff
@@ -684,15 +729,134 @@ func (k Keeper) GetByteCode(ctx sdk.Context, codeID uint64) ([]byte, error) {
+	defer func() {
+		if r := recover(); r != nil {
+			// if it's not an OutOfGas error, raise it again
+			// FIXME: do I need to do something special to get the proper stack trace here?
```
We have similar logic in the querier https://github.com/CosmWasm/wasmd/blob/master/x/wasm/internal/keeper/querier.go#L157-L175

ethanfrey replied:
Okay, that wants to catch all panics, I want to re-raise any other panic (just handle out of gas here).

alpe reviewed (commented):
Some quick feedback on the questions. No time for a proper review yet.

alpe commented on x/wasm/internal/keeper/keeper.go:
```diff
@@ -684,15 +729,134 @@ func (k Keeper) GetByteCode(ctx sdk.Context, codeID uint64) ([]byte, error) {
+				Ok: &wasmvmtypes.SubcallResponse{
+					Events: sdkEventsToWasmVmEvents(events),
+					// just take the first one for now if there are multiple sub-sdk messages
+					Data: data[0],
```
Req: `data` can be nil with `EncodeIBCMsg` for example 💣 
To be correct we should change the type to `[][]byte` for data. Another option would be to let the messenger "merge" the result data and return only a single element there. With both options I can see a communication issue for developers to use the data field proper without an understanding of the messenger implementation.

ethanfrey replied:
Nice find.

Tested and reproduced in https://github.com/CosmWasm/wasmd/pull/441/commits/c1d31ecca2b3a79429da9dc564c2984fd7e2f14d

Fixed in https://github.com/CosmWasm/wasmd/pull/441/commits/aba0596d720e6d2a818f528de0e6420422601491

alpe commented on x/wasm/internal/keeper/submsg_test.go:
```diff
@@ -0,0 +1,375 @@
+	}, sender.Attributes[0])
+
+	// where does this come from?
+	module := sub.Events[2]
```
The event-filter is built into the [handler](https://github.com/CosmWasm/wasmd/blob/master/x/wasm/handler.go#L51) currently. It guess we need it in the keeper now.

ethanfrey replied:
We don't really need it there, I guess it would just be cleaner?

Let's make another issue for event filtering. Or just investigate more during https://github.com/CosmWasm/wasmd/issues/445

alpe replied:
I will add a new issue for this.

alpe commented on x/wasm/internal/keeper/submsg_test.go:
```diff
@@ -0,0 +1,375 @@
+	checkAccount(t, ctx, accKeeper, bankKeeper, contractAddr, sdk.NewCoins(sdk.NewInt64Coin("denom", 25000)))
+	checkAccount(t, ctx, accKeeper, bankKeeper, creator, creatorBalance)
+
+	// TODO: query the reflect state to ensure the result was stored
```
Any todo left?

alpe commented on x/wasm/internal/keeper/submsg_test.go:
```diff
@@ -0,0 +1,375 @@
+			resultAssertions: []assertion{assertGasUsed(subGasLimit+51000, subGasLimit+53000), assertErrorString("out of gas")},
+		},
+
+		"instantiate contract gets address in data and events": {
```
good test cases 🌷

alpe reviewed (changes requested):
Good 🎸 
Only the data payload may become an issue.

ethanfrey commented:
Okay, I addressed all the comments and added a bug fix and test.

Rebased on master (to include pinned contract metering) and please feel free to merge if it looks better now.

alpe commented:
Awesome feature, Thanks 💐
